### PR TITLE
Code review only - not for checkin

### PR DESCRIPTION
**Pull request by dave-voltdb**
This is for a review of changes to date in the KO-integration branch.

### CODE REVIEW

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        DRPROD,
+    }
+
+    /*
+     * Output values. Package access intentionally; there is
+     * little point in adding a dozen or so simple-minded
+     * getters for internal-only use. Obviously these are
+     * named with prefixes similar to the Type names above.
+     */
+    long client_reqBytes, client_respMsgs, client_txns;
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:64]: I agree in general; in this case, however, I specifically want to make it clear the correspondance with the (input) type.  client_xY, client_fuBar, client_z convey this better than clientXY, clientFuBar, clientZ,

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                for (Pair<String,long[]> ent : ci.getLiveClientStats().values()) {
+                    long[] val = ent.getSecond(); // order of values is assumed
+                    reqBytes += val[1];
+                    respMsgs += val[2];
+                    txns += val[3];
+                }
+            }
+        }
+        catch (Exception ex) {
+            warn("checkClients", ex);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:127]: See the definition of 'warn' - "checkClients" is simply an insertion string, not a message.

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        DRCONS,
+        DRPROD,
+    }
+
+    /*
+     * Output values. Package access intentionally; there is
+     * little point in adding a dozen or so simple-minded
+     * getters for internal-only use. Obviously these are
+     * named with prefixes similar to the Type names above.
+     */
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:63]: I didn't want that.

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    }
+
+    /*
+     * Find source for someone else's stats (used for DR activity check)
+     */
+    private static StatsSource getStatsSource(StatsSelector selector) {
+        StatsSource ss = null;
+        StatsAgent sa = VoltDB.instance().getStatsAgent();
+        if (sa != null) {
+            Set<StatsSource> sss = sa.lookupStatsSource(selector, 0);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:277]: Yes

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    /**
+     * Status listener:
+     * @param intf  interface on which to listen for connections (null for any)
+     * @param port  TCP port number (zero for automatic)
+     * @param publicIntf  address to be returned in 'Host' header (if different)
+     */
+    public StatusListener(String intf, int port, String publicIntf) {
+        m_resolvedIntf = resolveInterface(intf, "");
+        m_publicIntf = resolveInterface(publicIntf, m_resolvedIntf);
+        initServer(m_resolvedIntf, port);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:74]: Pattern copied from HTTPAdminListener.

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            ctxHandler.setMaxFormKeys(MAXKEYS);
+            ctxHandler.addServlet(StatusServlet.class, "/status").setAsyncSupported(true);
+            server.setHandler(ctxHandler);
+        }
+
+        catch (Exception ex) {
+            logError("StatusListener: initialization failure: %s", ex);
+            try { connector.close(); } catch (Exception e2) { }
+            try { server.destroy(); } catch (Exception e2) { }
+            try { qtp.join(); } catch (Exception e2) { }
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:100]: Pattern copied from HTTPAdminListener.

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    private static final int MAXQUERY = 256;
+    private static final int MAXKEYS = 2;
+
+    private final String m_resolvedIntf;
+    private final String m_publicIntf;
+
+    private Server m_server;
+    private ServerConnector m_connector;
+    private String m_hostHeader;
+    private static StatusListener singleton;
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:62]: It's a singleton, I need to keep a reference somewhere, and I don't want to set it up before it's operational.

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    private static void logWarning(String str, Object... args) {
+        doLog(Level.WARN, str, args);
+    }
+
+    private static void logInfo(String str, Object... args) {
+        doLog(Level.INFO, str, args);
+    }
+
+    private static void logDebug(String str, Object... args) {
+        doLog(Level.DEBUG, str, args);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:271]: What you're saving there is the time spent formatting.  But this is once-only initialization.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:74]: Does not throw; initServer handles that.

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+     */
+    private static boolean isActive(String fmt, long... counts) {
+        boolean actv = false;
+        for (long count : counts) {
+            actv |= (count != 0);
+        }
+        if (actv) {
+            Long[] args = new Long[counts.length];
+            for (int i=0; i<counts.length; i++) {
+                args[i] = counts[i]; // boxing
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:306]: Didn't want to use Long... since I guess most values will be zero and I didn't want to box if I didn't have to. Micro-optimization...

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        }
+        return actv;
+    }
+
+    /*
+     * Standardized logging for exception handling. The exception
+     * is otherwise ignored.
+     */
+    private static void warn(String func, Exception ex) {
+        logger.warn(String.format("Unexpected exception in ActivityHelper.%s: %s", func, ex));
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:318]: ok, will do.

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+            if (ci != null) {
+                for (Pair<String,long[]> ent : ci.getLiveClientStats().values()) {
+                    long[] val = ent.getSecond(); // order of values is assumed
+                    reqBytes += val[1];
+                    respMsgs += val[2];
+                    txns += val[3];
+                }
+            }
+        }
+        catch (Exception ex) {
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:126]: Looks ugly.

File: src/frontend/org/voltdb/NodeStateTracker.java [hunk 12]
```diff
@@ -16,21 +16,97 @@
...
 
+import org.voltcore.logging.VoltLogger;
 import org.voltdb.common.NodeState;
 
 import com.google_voltpatches.common.base.Supplier;
 
 /**
- * Class that aides in the tracking of a VoltDB node state.
+ * Class that aids in the tracking of a VoltDB node state.
+ *
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/NodeStateTracker.java:29]: I'm ok with the blanks being removed in Javadoc but I still like them in the source.

File: src/frontend/org/voltdb/PauseActivityStats.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        return new ActivityHelper.OneShotIterator();
+    }
+
+    /*
+     * Main stats collection. We return a single row, and
+     * the key is irrelevant. Stats collected match those
+     * used by 'voltadmin pause --wait'.
+     */
+    private static final ActivityHelper.Type[] statsList = {
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/PauseActivityStats.java:79]: It's a "list" as defined in English, not a "List".

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 14]
```diff
@@ -961,6 +972,16 @@ public void initialize(Configuration config) {
                         consoleLog.info(String.format("Licensed to: %s", licensee));
                     }
                 }
+
+                // Start the listener that responds to "status" requests. This needs
+                // to be started "early" to be available during initialization.
+                StatusListener sl = new StatusListener(config.m_statusInterface,
+                                                       config.m_statusPort,
+                                                       config.m_publicInterface);
+                sl.start();
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java]: Yup.  There are calls to StatusListener.shutdown() at appropriate places. It's a singleton.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 15]
```diff
@@ -961,6 +972,16 @@ public void initialize(Configuration config) {
...
+                config.m_statusInterface = sl.getListenInterface();
+                config.m_statusPort = sl.getAssignedPort();
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java]: That's what every other port qualifier does.  See HTTPAdminListener, which is where I got it from (I think).

File: src/frontend/org/voltdb/ActivityHelper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:64]: changed

File: src/frontend/org/voltdb/ActivityHelper.java [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+     * DR producer.
+     * Returns details of outstanding data, expressed in bytes and message
+     * segments, summed across all partitions. Zero when none outstanding.
+     */
+    private boolean checkDrProducer() {
+        long bytesPend = 0, segsPend = 0;
+        try {
+            StatsSource ss = getStatsSource(StatsSelector.DRPRODUCERPARTITION);
+            if (ss != null) {
+                int ix_totalBytes = getIndex(ss, DRProducerStatsBase.Columns.TOTAL_BYTES);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ActivityHelper.java:213]: done

File: src/frontend/org/voltdb/CommandLog.java [hunk 17]
```diff
@@ -201,6 +201,14 @@ public abstract void initForRejoin(int logSize,
      */
     public void populateCommandLogStats(Map<String, Integer> columnNameToIndex, Object[] rowValues);
 
+    /**
+     * Statistics-related interface, used by ActivityStats.
+     * Implementaton should return outstanding byte count
+     * in out[0], outstanding txn count in out[1].
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/CommandLog.java:207]: I just want two numbers returned without adding a lot of mechanism.  An array is fine.

File: src/frontend/org/voltdb/NodeStateTracker.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/NodeStateTracker.java:29]: added.  sigh.

File: src/frontend/org/voltdb/PauseActivityStats.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+    /*
+     * Constructs a description of the columns we'll return
+     * in our stats table.
+     */
+    @Override
+    protected void populateColumnSchema(ArrayList<ColumnInfo> columns) {
+        super.populateColumnSchema(columns);
+        for (ColumnName col : ColumnName.values()) {
+            VoltType type = (col == ColumnName.ACTIVE ? VoltType.TINYINT : VoltType.BIGINT);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/PauseActivityStats.java:60]: Seemed like more mechanism and more flexibility than was actually necessary for such a trivial thing.  YAGNI.

File: src/frontend/org/voltdb/RealVoltDB.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java]: Yup. HTTPAdminListener, in module Inits. 

The config may not contain the actual interface/port if it was defaulted. There is also find-a-free-port logic which I copied, so the port is determined at run time (seems a bit odd but...).

According to comments for HTTPAdmin, there's code that expects to use the actual port from the config.

[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java]: Leaving as-is

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 19]
```diff
@@ -1482,6 +1502,10 @@ public void initialize(Configuration config) {
             // Dummy DRCONSUMER stats
             replaceDRConsumerStatsWithDummy();
 
+            // Operator function helpers
+            getStatsAgent().registerStatsSource(StatsSelector.SHUTDOWN_CHECK, 0, new ShutdownActivityStats());
+            getStatsAgent().registerStatsSource(StatsSelector.PAUSE_CHECK, 0, new PauseActivityStats());
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java:1509]: I agree with the sentiment, but I like consistency in code.

```
$ fgrep 'getStatsAgent().register' RealVoltDB.java  | wc -l
21

```

File: src/frontend/org/voltdb/PauseActivityStats.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+ * relates to, and perhaps whether forward progress is being made.
+ */
+public class PauseActivityStats extends StatsSource
+{
+    private static final VoltLogger logger = new VoltLogger("HOST");
+
+    private enum ColumnName {
+        ACTIVE, // 0 if all other gauges 0, else 1
+        EXPORTS_PENDING,
+        DRPROD_SEGS, DRPROD_BYTES,
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/PauseActivityStats.java:45]: segments.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 21]
```diff
@@ -4241,7 +4274,7 @@ public void run() {
                         System.exit(0);
                     }
                     try {
-                        Thread.sleep(5);
+                        Thread.sleep(5); // 5 mSec
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java:4279]: Of course it's the API, but I had to look it up to be sure; I wanted to leave a note to that effect.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 22]
```diff
@@ -4378,19 +4412,18 @@ private void onRejoinCompletion() {
 
                 String actionName = m_joining ? "join" : "rejoin";
                 m_joining = false;
-                consoleLog.info(String.format("Node %s completed", actionName));
+                consoleLog.info(String.format("Node %s completed", actionName)); // onRejoinCompletion
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java:4417]: The reader knows when he's explicitly reading that method, but when searching around to find the places at which startup was completed, it was helpful to know which one I was looking at.

I'll leave it there unless you really object.

RealVoltDB.java:                consoleLog.info(String.format("Node %s completed", actionName)); // onRejoinCompletion
RealVoltDB.java:                consoleLog.info(String.format("Node %s completed", actionName)); // onRecoveryComplete
gr

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 23]
```diff
@@ -4437,6 +4470,42 @@ public OperationMode getStartMode()
         return m_startMode;
     }
 
+    @Override
+    public NodeState getNodeState()
+    {
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java:4477]: My cardinal rule is "when adding code, use the existing style".  I don't like it either, but the method I added is in the middle of an expanse of code using "C-style" brace placement.

File: src/frontend/org/voltdb/ShutdownActivityStats.java [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+                active = helper.collect(statsWithCmdLog);
+                setValue(row, ColumnName.CLIENT_TXNS, helper.client_txns);
+                setValue(row, ColumnName.CLIENT_REQ_BYTES, helper.client_reqBytes);
+                setValue(row, ColumnName.CLIENT_RESP_MSGS, helper.client_respMsgs);
+                setValue(row, ColumnName.IMPORTS_PENDING, helper.import_pend);
+                setValue(row, ColumnName.CMDLOG_TXNS, helper.cmdlog_txns);
+                setValue(row, ColumnName.CMDLOG_BYTES, helper.cmdlog_bytes);
+                setValue(row, ColumnName.EXPORTS_PENDING, helper.export_pend);
+            }
+            else {
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/ShutdownActivityStats.java:124]: Now there I differ.  The eye is more readily drawn to the keyword at the beginning of the line. I am not one to carelessly squander vertical space, but this is worth it.

File: src/frontend/org/voltdb/utils/StatusServlet.java [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+        }
+    }
+
+    private String progress(VoltDBInterface instance) {
+        String p = "complete";
+        if (!instance.getNodeStartupComplete()) {
+            int[] n = instance.getNodeStartupProgress();
+            if (n[1] > 0) {
+                if (n[0] < 0) n[0] = 0;
+                if (n[0] > n[1]) n[0] = n[1];
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusServlet.java:77]: Seems trivial but ok.

File: src/frontend/org/voltdb/utils/StatusServlet.java [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+            throw new ServletException(ex);
+        }
+    }
+
+    private String progress(VoltDBInterface instance) {
+        String p = "complete";
+        if (!instance.getNodeStartupComplete()) {
+            int[] n = instance.getNodeStartupProgress();
+            if (n[1] > 0) {
+                if (n[0] < 0) n[0] = 0;
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusServlet.java:76]: Could, but that results in assigning n[0] to n[0] most of the time.  And it loses the relationship to the next line.

File: src/frontend/org/voltdb/VoltDBInterface.java [hunk 27]
```diff
@@ -243,6 +244,20 @@ default public void cleanUpTempCatalogJar()
 
     public OperationMode getStartMode();
 
+    public StartAction getStartAction();
+
+    public NodeState getNodeState();
+
+    public boolean getNodeStartupComplete();
+
+    public int[] getNodeStartupProgress();
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/VoltDBInterface.java:253]: I'm a fan of simple structures for simple uses, such as returning two numbers.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:100]: Leaving as-is for similarity with existing code.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:62]: The specific problem is getting from an executing jetty servlet back to the listener.  Somewhere there's got to be a static reference.  I could have put a non-static reference in RealVoltDB, which itself is found through a static reference, and then made a member getStatusListener(), but that pointlessly extends what amounts to a private arrangement between the listener and the servlets it manages. i.e., this way has less spaghetti.  It would have been nice if the jetty programmers had provided a way to pass suitable context around, but they did not.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:271]: One use only - failure to open port in the auto-assign-port-number mode. Leaving as is.

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 28]
```diff
@@ -0,0 +1,273 @@
...
+    private static String resolveInterface(String intf, String defIntf) {
+        String temp = (intf == null ? "" : intf.trim());
+        return temp.isEmpty() ? defIntf : temp;
+    }
+
+    private static void doLog(Level level, String str, Object[] args) {
+        if (args.length != 0) {
+            str = String.format(str, args);
+        }
+        m_log.log(level, str, null);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:255]: done

File: src/frontend/org/voltdb/PauseActivityStats.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/PauseActivityStats.java:45]: It appears I made up the word "segment" for the quantity of data represented by a DrId.  Is there a better (short) word I could use for a column name?

                        segsPend += lastQueuedDrId - lastAckedDrId;

I don't really care what they are; this is just a number that indicates in some vague way that there's work to do, and the number should be trending down.  But I have to have a column name!

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 29]
```diff
@@ -3747,6 +3773,10 @@ public boolean shutdown(Thread mainSiteThread) throws InterruptedException {
                 org.voltdb.iv2.InitiatorMailbox.m_allInitiatorMailboxes.clear();
 
                 PartitionDRGateway.m_partitionDRGateways = ImmutableMap.of();
+
+                // We left the status API up as long as possible...
+                StatusListener.shutdown();
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/RealVoltDB.java:3780]: There is one static ref, and it is essential by virtue of jetty.  Since I have to have it for jetty, there's little downside in using it for shutdown convenience.

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 30]
```diff
@@ -0,0 +1,273 @@
...
+            }
+        }
+        String fail;
+        if (portLo == portHi) {
+            fail = String.format("Unable to open port %s", portLo);
+        }
+        else {
+            fail = String.format("Unable to open any port in range %s to %s", portLo, portHi);
+        }
+        throw new RuntimeException(fail);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:135]: I disagree since that involves then modifying the signatures of a stack of routines.

Checked exceptions are widely regarded, including by reputable sources, as a failed experiment. This is a relatively sober discussion: https://www.ibm.com/developerworks/library/j-jtp05254/index.html

My view is expressed in 'unstable method signatures'.

File: src/frontend/org/voltdb/utils/StatusListener.java [hunk 31]
```diff
@@ -0,0 +1,273 @@
...
+            ctxHandler.addServlet(StatusServlet.class, "/status").setAsyncSupported(true);
+            server.setHandler(ctxHandler);
+        }
+
+        catch (Exception ex) {
+            logError("StatusListener: initialization failure: %s", ex);
+            try { connector.close(); } catch (Exception e2) { }
+            try { server.destroy(); } catch (Exception e2) { }
+            try { qtp.join(); } catch (Exception e2) { }
+            throw new RuntimeException("Failed to initialize status listener", ex);
```
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:101]: as before

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:74]: Beg pardon; does throw in fatal initialization error  Non-cleanup is not an issue in that case.

File: src/frontend/org/voltdb/PauseActivityStats.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/PauseActivityStats.java]: Is now 'ROWS' based on conversation with Nate & Manju.

Doc is undecided about rows and tuples; it uses both. I think rows is plainer.

File: src/frontend/org/voltdb/ActivityHelper.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (n3world) [reply src/frontend/org/voltdb/ActivityHelper.java:307]: Auto boxing of any number between -128 and 127 is cached by java and essentially a static which is reused so autoboxing of values in this range is cheap

File: src/frontend/org/voltdb/RealVoltDB.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (n3world) [reply src/frontend/org/voltdb/RealVoltDB.java:1509]: I will quote my wife when talking about something one of her colleagues did "Following a pattern just to follow a pattern is not a good excuse for bad coding"

File: src/frontend/org/voltdb/RealVoltDB.java (hunk 29 above)
[the remark this answers is not in the document]
Comment (n3world) [reply src/frontend/org/voltdb/RealVoltDB.java:3780]: Statics often cause lots of problems in unit tests when the test doesn't know that something dangling over there could be tripped over by the next test.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 30 above)
[the remark this answers is not in the document]
Comment (n3world) [reply src/frontend/org/voltdb/utils/StatusListener.java:135]: I disagree with this from experience. When code throws runtime exceptions nobody catches them correctly because they are never documented. Checked exceptions force the documentation and force the implementer to consider how different error conditions should be handled. This has changed more recently but it was generally considered a bad thing to catch Exception or anything generic like that instead you should catch the specific exception for the error case you are trying to handle.

Basically almost all of the time a method which throws runtime exceptions will not document it so they will not be caught and this will only be noticed when some bug is hit at a customer site and the whole system crashes because a caller did not know this exception should be handled.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (n3world) [reply src/frontend/org/voltdb/utils/StatusListener.java:271]: Yes you saving parsing a string splitting it into multiple strings and piecing it back together with other variables which are converted from an object to a string. When this extra work is done once during initialization it isn't a big deal but this was more of a in general you should surround debug log statements which do string formatting with a check if debug is enabled.

File: src/frontend/org/voltdb/utils/StatusListener.java (hunk 30 above)
[the remark this answers is not in the document]
Comment (n3world) [reply src/frontend/org/voltdb/utils/StatusListener.java:135]: BTW from the article you posted

> Document, document, document
> 
> If you do decide to use unchecked exceptions, you need to document this choice thoroughly, including documenting in Javadoc all the unchecked exceptions a method might throw. Johnson suggests making the decision between checked and unchecked exceptions on a package-by-package basis. When using unchecked exceptions, also remember that you may need to use try...finally blocked even when you are not catching any exceptions, so that you can perform cleanup actions such as closing database connections. With checked exceptions, we have try...catch to remind us to add a finally clause. With unchecked exceptions, we don't have that crutch to lean on.

[the remark this answers is not in the document]
Comment (dave-voltdb) [reply src/frontend/org/voltdb/utils/StatusListener.java:135]: My position too is from experience.

The system is MEANT to not come up if this API cannot initialize.

**Comment by dave-voltdb** [comment]
Closing this PR - review complete
